### PR TITLE
Fix crash when dictionaries are deallocated during logEvent

### PR DESCRIPTION
**Pull request by djih**
Will release a v2.5.1 off of this branch, and then apply same changes to master for v3.0.1

### CODE REVIEW

**Comment by curtisliu** [comment]
You mean v2.5.2? :)

**Comment by curtisliu** [comment]
I think https://github.com/amplitude/Amplitude-iOS/pull/39/files#diff-2777663457a583bbf669698eaadec79bR999 should also be autoreleased

File: Amplitude/Amplitude.m [hunk 1]
```diff
@@ -378,7 +378,9 @@ - (void)logEvent:(NSString*) eventType withEventProperties:(NSDictionary*) event
     if (timestamp == nil) {
         timestamp = [NSNumber numberWithLongLong:[[NSDate date] timeIntervalSince1970] * 1000];
     }
-    
+
+    eventProperties = [eventProperties copy];
+    apiProperties = [apiProperties mutableCopy];
```
Comment (curtisliu) [inline Amplitude/Amplitude.m]: want to do this with userProperties as well?

File: Amplitude/Amplitude.m [hunk 2]
```diff
@@ -401,7 +403,9 @@ - (void)logEvent:(NSString*) eventType withEventProperties:(NSDictionary*) event
             [event setValue:eventType forKey:@"event_type"];
             [event setValue:[NSNumber numberWithLongLong:newId] forKey:@"event_id"];
             [event setValue:[self replaceWithEmptyJSON:eventProperties] forKey:@"event_properties"];
+            (void) SAFE_ARC_AUTORELEASE(eventProperties);
```
Comment (curtisliu) [inline Amplitude/Amplitude.m]: i think the AUTORELEASES should happen together for better code organization. Also needs to happen if the block returns before this.

File: Amplitude/Amplitude.m (hunk 1 above)
Comment (djih) [inline Amplitude/Amplitude.m]: Added

File: Amplitude/Amplitude.m (hunk 2 above)
Comment (djih) [inline Amplitude/Amplitude.m]: Ah good catch

**Comment by djih** [comment]
Copy user properties, released objCopy, PTAL @curtisliu

File: Amplitude/Amplitude.m [hunk 3]
```diff
@@ -402,7 +409,11 @@ - (void)logEvent:(NSString*) eventType withEventProperties:(NSDictionary*) event
             [event setValue:[NSNumber numberWithLongLong:newId] forKey:@"event_id"];
             [event setValue:[self replaceWithEmptyJSON:eventProperties] forKey:@"event_properties"];
             [event setValue:[self replaceWithEmptyJSON:apiProperties] forKey:@"api_properties"];
-            [event setValue:[self replaceWithEmptyJSON:_userProperties] forKey:@"user_properties"];
+            [event setValue:[self replaceWithEmptyJSON:userProperties] forKey:@"user_properties"];
+
+            (void) SAFE_ARC_AUTORELEASE(eventProperties);
```
Comment (curtisliu) [inline Amplitude/Amplitude.m]: should these be SAFE_ARC_RELEASE?

**Comment by curtisliu** [comment]
LGTM

**Pull request closed without merging**